README.md

```diff
@@ -361,6 +361,29 @@ To build standalone source code package, the following commands can be issued:
 
 After that, a stanalone archive with source code will be created in the `.build` directory.
 
+When cross compiling, the AS, AR, CC, CXX, LD, etc. variables should be set according to the target/cross compile toolchain. Additional variables should be configured:
+  * set the ARCHITECTURE option to the target architecture
+  * set the CROSS_COMPILE option to 1
+  * set HOST_* options for the build host (HOST_AS, HOST_AR, HOST_CC, HOST_CXX, HOST_LD, HOST_CFLAGS, HOST_CXXFLAGS, HOST_EXE_FLAGS, HOST_SO_FLAGS)
+  * set the PREFIX option to a non-system directory
+
+Example cross compile procedure for aarch64 target on x86-64 build host:
+```
+  make clean
+  make config ARCHITECTURE="aarch64" CROSS_COMPILE="1" \
+    HOST_AS="as" HOST_AR="ar" HOST_CC="gcc" HOST_CXX="g++" HOST_LD="ld" \
+    HOST_CFLAGS="-O2 -pipe -g -feliminate-unused-debug-types -march=x86-64 -fvisibility=hidden -fdata-sections -ffunction-sections -fno-asynchronous-unwind-tables -pipe -Wall" \
```

Comment:
It looks like the primary parameter here is `-march` and other ones repeat the `CFLAGS`. Maybe this can be automatically calculated in makefiles when `CROSS_COMPILE` is set to 1.

Reply:
Yes, I think that if CROSS_COMPILE=1, you can default HOST_AS=as, HOST_AR=ar, HOST_CC=gcc, etc. instead of defaulting them to HOST_AS=$(AS), HOST_AR=$(AR), HOST_CC=$(CC) ... That would simplify things.

I guess I'm not exactly sure what to do about the HOST_*FLAGS variables though. Not sure how to get reasonable values from the build host.

Reply:
You can just paste somewhere in makefiles insturctions like this:
`$(info AS=$(AS))`.
I suppose that `HOST_CXX` and `CXX` are still similar and there is no need to redefine them.
By the other side, the HOST_CFLAGS is important because it holds the architecture. But if we compile source code for host platform, we probably can just omit them.

Reply:
I'm not sure that I understand. Currently `HOST_CXX` must be redefined because if it's not set then the Makefile sets it to `$(CXX)` ([`HOST_CXX ?= $(CXX)`](https://github.com/lsp-plugins/lsp-plugin-fw/blob/master/make/tools.mk#L53)), which is the cross compiler not the native compiler.

If CROSS_COMPILE=1, the Makefile can (should?) assume that CXX is the cross compiler (something like `aarch64-linux-g++`) and it can default `HOST_CXX ?= g++` (which is typically a symlink to the native compiler- `/usr/bin/x86_64-linux-gnu-g++-10` on my Debian 11 system) to set the compiler to the native build host binary instead of the cross compiler binary.

Reply:
I mean, you can add a code to makefile that will dump the desired variables to standard output using the `$(info ...)` function. This could be very useful to understand actual values assigned to variables in your case.
The first question is, do we really need to redefine `HOST_CXX` and other variables? Maybe this is not required.
Second, we need to understand the build flags. Maybe we can automate them at the configure stage.

The main target here is, I want to get as less as possible additional options passed to the `make config` command to make cross-compilation easier.

Reply:
I guess this is the dilemma. You can't really assume that `HOST_CXX` should be set to  `g++`, but I don't know of a good way to get the correct `HOST_CXX` value in my cross compile environment since `CXX` is set to the cross compiler.

Reply:
I get it- I want to pass fewer options to `make config` too. The shell environment in which I'm building lsp-plugins (a yocto generated cross compile SDK, but similar to other cross compile toolchains) has all of these variables overridden in order to point to the cross compiler binaries:
```
CPP=aarch64-poky-linux-gcc
CXX=aarch64-poky-linux-g++
LD=aarch64-poky-linux-ld
...
```
so any attempt to get the value of `$(CXX)` will just give you back the cross compiler binary.

But I didn't see those [`X_<z>_TOOL` variables](https://github.com/lsp-plugins/lsp-plugin-fw/blob/master/make/tools.mk#L23) earlier. Maybe it's as simple as defaulting all of the [`HOST_*` variables](https://github.com/lsp-plugins/lsp-plugin-fw/blob/master/make/tools.mk#L51) to the `X_<z>_TOOL` values?
```
# Define tool variables for host build                                                                                                                                               
HOST_CC            ?= $(X_CC_TOOL)
HOST_CXX           ?= $(X_CXX_TOOL)
HOST_AS            ?= $(X_AS_TOOL)
HOST_AR            ?= $(X_AR_TOOL)
HOST_LD            ?= $(X_LD_TOOL)
HOST_PHP           ?= $(X_PHP_TOOL)
HOST_PKG_CONFIG    ?= $(X_PKG_CONFIG_TOOL)
```

That would allow my environment to override `CXX` without affecting `HOST_CXX`...?

Reply:
Yes, I think this is a good idea of defaulting to these values instead of `$(CC)`, etc. I think it is a good first step to improve the cross-compilation.